gluten-core/src/main/scala/org/apache/gluten/extension/columnar/OffloadSingleNode.scala

```diff
@@ -181,7 +185,135 @@ case class OffloadJoin() extends OffloadSingleNode with LogLevelUtil {
+      case p @ ProjectExec(projectList, child) =>
+        p.copy(genNewProjectList(projectList), addMetadataCol(child, replacedExprs))
+      case p @ ProjectExecTransformer(projectList, child) =>
+        p.copy(genNewProjectList(projectList), addMetadataCol(child, replacedExprs))
+      case _ => plan.withNewChildren(plan.children.map(addMetadataCol(_, replacedExprs)))
+    }
+  }
+
+  private def tryOffloadProjectExecWithInputFileRelatedExprs(
+      projectExec: ProjectExec): SparkPlan = {
+    def findScanNodes(plan: SparkPlan): Seq[SparkPlan] = {
+      plan.collect {
+        case f @ (_: FileSourceScanExec | _: AbstractFileSourceScanExec |
+            _: DataSourceV2ScanExecBase) =>
+          f
+      }
+    }
+    val addHint = AddTransformHintRule()
+    val newProjectList = projectExec.projectList.filterNot(containsInputFileRelatedExpr)
+    val newProjectExec = ProjectExec(newProjectList, projectExec.child)
+    addHint.apply(newProjectExec)
+    if (TransformHints.isNotTransformable(newProjectExec)) {
+      // Project is still not transformable after remove `input_file_name` expressions.
+      projectExec
+    } else {
+      // the project with `input_file_name` expression should have at most
+      // one data source, reference:
+      // https://github.com/apache/spark/blob/e459674127e7b21e2767cc62d10ea6f1f941936c
+      // /sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/rules.scala#L506
+      val leafScans = findScanNodes(projectExec)
+      assert(leafScans.size <= 1)
+      if (leafScans.isEmpty || TransformHints.isNotTransformable(leafScans(0))) {
+        // It means
+        // 1. projectExec has `input_file_name` but no scan child.
+        // 2. It has scan child node but the scan node fallback.
+        projectExec
+      } else {
+        val replacedExprs = scala.collection.mutable.Map[String, AttributeReference]()
+        val newProjectList = projectExec.projectList.map {
+          expr => rewriteExpr(expr, replacedExprs).asInstanceOf[NamedExpression]
+        }
+        val newChild = addMetadataCol(projectExec.child, replacedExprs)
+        logDebug(
+          s"Columnar Processing for ${projectExec.getClass} with " +
+            s"ProjectList ${projectExec.projectList} is currently supported.")
+        ProjectExecTransformer(newProjectList, newChild)
+      }
+    }
+  }
+
+  private def genProjectExec(projectExec: ProjectExec): SparkPlan = {
+    if (!TransformHints.isNotTransformable(projectExec)) {
+      logDebug(s"Columnar Processing for ${projectExec.getClass} is currently supported.")
+      ProjectExecTransformer(projectExec.projectList, projectExec.child)
+    } else if (
+      BackendsApiManager.getSettings.supportNativeInputFileRelatedExpr() &&
+      projectExec.projectList.exists(containsInputFileRelatedExpr)
+    ) {
+      tryOffloadProjectExecWithInputFileRelatedExprs(projectExec)
+    } else projectExec
```

Comment:
Can we invert these if conditions?

e.g.

```
if (not-transformable and has input_file_name) then ... elif (not-transformable) then ... else ...
```

Reply:
If invertable, let's also avoid using `else`s to make code clearer.

Reply:
I invert the condition but don't actually know if that is you want since don't think it can make code more clearer.